crates/ruff_python_formatter/src/comments/placement.rs

```diff
@@ -1233,6 +1234,124 @@ fn find_only_token_str_in_range(range: TextRange, locator: &Locator, token_str:
+        let mut tokens =
+            SimpleTokenizer::new(locator.contents(), TextRange::new(start, end)).skip_trivia();
+        tokens.next().map(|t| t.start())
+    }
+
+    let AnyNodeRef::Comprehension(comprehension) = comment.enclosing_node() else {
+        return CommentPlacement::Default(comment);
+    };
+    let is_own_line = comment.line_position().is_own_line();
+
+    if comment.slice().end() < comprehension.target.range().start() {
+        return if is_own_line {
+            // own line comments are correctly assigned as leading the target
+            CommentPlacement::Default(comment)
+        } else {
+            // after the `for`
+            CommentPlacement::dangling(comment.enclosing_node(), comment)
+        };
+    }
+
+    let mut tokens = SimpleTokenizer::new(
+        locator.contents(),
+        TextRange::new(
+            comprehension.target.range().end(),
+            comprehension.iter.range().start(),
+        ),
+    )
+    .skip_trivia();
+    let Some(in_token) = tokens.next() else {
+        // Should always have an `in`
+        debug_assert!(false);
+        return CommentPlacement::Default(comment);
+    };
+
+    if comment.slice().start() < in_token.start() {
+        // attach as dangling comments on the target
+        // (to be rendered as leading on the "in")
+        return if is_own_line {
+            CommentPlacement::dangling(comment.enclosing_node(), comment)
+        } else {
+            // correctly trailing on the target
+            CommentPlacement::Default(comment)
+        };
+    }
+
+    if comment.slice().start() < comprehension.iter.range().start() {
+        // attach as dangling comments on the iter
+
+        return if is_own_line {
+            // after the `in` and own line: leading on the iter
+            CommentPlacement::leading((&comprehension.iter).into(), comment)
+        } else {
+            // after the `in` but same line, turn into trailin on the `in` token
+            CommentPlacement::dangling((&comprehension.iter).into(), comment)
+        };
+    }
+
+    let mut last_end = comprehension.iter.range().end();
+
+    for if_node in &comprehension.ifs {
+        // [
+        //     a
+        //     for
+        //     c
+        //     in
+        //     e
+        //     # above if   <-- find these own-line between previous and `if` token
+        //     if  # if     <-- find these end-of-line between `if` and if node (`f`)
+        //     # above f    <-- already correctly assigned as leading `f`
+        //     f  # f       <-- already correctly assigned as trailing `f`
+        //     # above if2
+        //     if  # if2
+        //     # above g
+        //     g  # g
+        // ]
+        let Some(if_token_start) = find_if_token(locator, last_end, if_node.range().end()) else {
+            // there should always be an `if` here
+            debug_assert!(false);
+            return CommentPlacement::Default(comment);
+        };
```

Comment:
Nit: We should assert that this is indeed the expected `if` token and not any other token. You may also need to deal with any `(` or `)` in case the expression can be parenthesized.